Sources/NIOHPACK/HeaderTables.swift

```diff
@@ -0,0 +1,324 @@
+        
+        var newLength = self.length + entry.length
+        if newLength > self.maxSize {
+            self.purge(toRelease: newLength - maxSize)
+            newLength = self.length + entry.length
+            
+            if newLength > self.maxSize {
+                // Danger, Will Robinson! We can't free up enough space!
+                throw RingBufferError.BufferOverrun(amount: newLength - self.maxSize)
+            }
+        }
+        
+        self.headers.prepend(entry)
+        self.length = newLength
+    }
+    
+    /// Purges `toRelease` bytes from the table, where 'bytes' refers to the byte-count
+    /// of a table entry specified in RFC 7541: [name octets] + [value octets] + 32.
+    ///
+    /// - parameter toRelease: The table entry length of bytes to remove from the table.
+    mutating func purge(toRelease count: Int) {
+        guard count < self.length else {
+            // clear all the things
+            self.headers.removeAll()
+            self.buffer.clear()
+            self.length = 0
+            return
+        }
+        
+        var available = self.maxSize - self.length
+        let needed = available + count
+        while available < needed && !self.headers.isEmpty {
+            available += self.purgeOne()
+        }
+    }
+    
+    @discardableResult
+    private mutating func purgeOne() -> Int {
+        precondition(self.headers.isEmpty == false, "should not call purgeOne() unless we have something to purge")
+        // Remember: we're removing from the *end* of the header list, since we *prepend* new items there, but we're
+        // removing bytes from the *start* of the storage, because we *append* there.
+        let entry = self.headers.removeLast()
+        self.buffer.moveHead(forwardBy: entry.name.length + entry.value.length)
+        self.length -= entry.length
+        return entry.length
+    }
+    
+    // internal for testing
+    func dumpHeaders(offsetBy amount: Int = 0) -> String {
+        return self.headers.enumerated().reduce("") {
+            $0 + "\($1.0 + amount) - \(self.string(idx: $1.1.name)) : \(self.string(idx: $1.1.value))\n"
+        }
+    }
+}
+
+extension HeaderTableStorage : CustomStringConvertible {
+    var description: String {
+        var array: [(String, String)] = []
+        for header in self.headers {
+            array.append((self.string(idx: header.name), self.string(idx: header.value)))
+        }
+        return array.description
+    }
+}
+
+private extension StringRing {
+    func equalCaseInsensitiveASCII<C : Collection>(view: C, at index: HPACKHeaderIndex) -> Bool where C.Element == UInt8 {
+        guard view.count == index.length else {
+            return false
+        }
+        return withVeryUnsafeBytes { buffer in
+            // This should never happens as we control when this is called. Adding an assert to ensure this.
+            let address = buffer.baseAddress!.assumingMemoryBound(to: UInt8.self)
```

Comment:
I don't think we need to do any binding assumptions here. `buffer` does just fine to produce `UInt8`s.